### ISSUE TITLE
Revision woes

### ISSUE THREAD

**Issue by mjpost** [acl-org/acl-anthology#904]
There is a lot of complexity and consequent room for user error in the current revisioning process:

- The unadorned PDF (`{anthid}.pdf`) is always replaced so as to work with our redirect rules
- For the first revision, `{anthid}v1.pdf` has to be created
- We have to update checksums on both the `<revision>` and `<url>` tags
- It's also strange how Revision 1 gets a "v2" extension

I wonder if we should get rid of `<revision>` and `<erratum>` in favor of generic `<url>` tags with a revision attribute. It would also be nice to avoid renaming at all. This would break the ".pdf" URL shortcut, but maybe that doesn't matter—users would just have to visit the paper page to get the latest revision.

> _Originally posted by @mjpost in https://github.com/acl-org/acl-anthology/pull/903#issuecomment-655524448_

**Comment by akoehn** [comment]
>I wonder if we should get rid of <revision> and <erratum> in favor of generic <url> tags with a revision attribute.

No: We then would need a `<url type="revision">` and `<url type="erratum">` because they behave differently.  A revision is the new default for the PDF button, an erratum is not.  IMO we would not gain anything by that change.

**Comment by mbollmann** [comment]
What part of the process leaves room for user error that couldn't be addressed by having a single script run through all these steps?

(NB: I don't know exactly what the script you currently use are doing.)

**Comment by mjpost** [comment]
It's `bin/add_revision.py`, which

- Downloads the existing PDF, renames to "v1"
- Downloads the new PDF, renames to "v2", overwrites the unadorned
- Puts them all locally where they can be synced up

Today (#903) I made edits to make this work with new-style IDs, and ended up pulling the HTML page and saving it as "v1.pdf". This is fixed now (and I added some sanity checks), but in general it just seems like bad practice to be overwriting files at all. I would have lost them if I didn't have a backup.

Re: the `<url>` tag, it's a little unseemly to have both it and `<revision>` pointing to the PDF. Do you disagree? It'd be sweeping change (a possible point against it), but it makes more sense to me to have a `<version rev="1">` tag (or something of that sort), whose absence would mean no PDF is available. We would keep `<erratum>` as is.

**Comment by mbollmann** [comment]
I have a deja-vu here: https://github.com/acl-org/acl-anthology/issues/295#issuecomment-624684119

**Comment by mjpost** [comment]
Ah, I thought this felt familiar. Okay, we stick with overwriting the PDF, and I will be more careful.

That leaves the tag discussion. Do you disagree that having `<url>` and `<revision>` (once there is one) is redundant?

I wonder if changing `<url>` to an explicit `<pdf revid="1">` link would be more clear. In this scheme, revisions would just add additional `<pdf>` tags. `<erratum>` would stay the same. This might also help clarify [a minor point of confusion](https://github.com/acl-org/acl-anthology/issues/823) between the notion of URL and PDF.

**Comment by mbollmann** [comment]
I agree it's not super elegant at the moment, but not fully redundant either. Assume there's one revision, then right now we produce links to `[id].pdf`, `[id]v1.pdf`, and `[id]v2.pdf`. Now, `[id].pdf` and whatever the latest revision is will always be identical, but we do always want the `[id].pdf` link. So in that sense, if anything, it's the `[id]v2.pdf` that's redundant.

So what you _could_ do is something like:

```
<pdf revid="1" file="2020.test-test.42v1" />
<pdf revid="2" file="2020.test-test.42">This is a revision because of xyz.</pdf>
```

and have the website always link the entry with the highest `revid` to the big PDF button. But note that there wouldn't be a `v2` anymore in this version, and if we added another revision, we'd have to rename `2020.test-test.42` to `2020.test-test.42v2`, so if that's really more intuitive ... I don't know.

**Comment by nschneid** [comment]
Can this be closed?

**Comment by mbollmann** [comment]
The redundancy in the XML remains, but I can’t think of a variant that I like better. 

`<pdf revision="2" ...>` would absolutely work, but we still want to generate the `[id].pdf` link (without a version number) somehow, and we also attach an explanation to revisions which doesn’t feel right within the semantics of a general `<pdf>` tag.

So, I don’t know if we want to close this or not, but I don’t have a better suggestion in any case.

**Comment by nschneid** [comment]
Summary of the concern: it's a little weird that we do musical chairs of filenames when a revision comes in, and there's redundancy in the XML. But an advantage is that search engines see the fixed canonical URL which is the most recent PDF. What does arXiv do?

**Comment by mjpost** [comment]
Agree with @mbollmann on the uncertainty. We could use `<pdf revision="2">Explanation</pdf>` but I think this is low-priority at the moment.

Updating for our current system:
- I dislike the file shuffling, it feels like the wrong thing to do (orig → v1, new → orig, new → v2)
- I dislike the redundancy: `<pdf>` and also the Nth revision are separate tags

**Comment by mbollmann** [comment]
The first version you upload to arXiv is v1 by definition. Appending "v1" to the URL of a freshly uploaded arXiv paper works and is identical to accessing the main article. (e.g. https://arxiv.org/abs/2609.22091 vs. https://arxiv.org/abs/2609.22091v1)

The solution to the file shuffling issue would be suffixing everything with "v1" by default, and make the base URL (without a version suffix) resolve to the latest version. That would probably involve creating symlinks for every paper, unless there’s some .htaccess magic that could achieve this.
